File: src/GCov/LineMarker/CoverageLineMarker.java
package GCov.LineMarker;

import GCov.Data.GCovCoverageGatherer;
import GCov.State.EditorState;
import com.intellij.codeInsight.daemon.LineMarkerInfo;
import com.intellij.codeInsight.daemon.LineMarkerProvider;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.markup.GutterIconRenderer;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.FileViewProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import icons.CoverageIcons;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;
import java.util.List;

public class CoverageLineMarker implements LineMarkerProvider {

    @Override
    public @Nullable LineMarkerInfo getLineMarkerInfo(@NotNull PsiElement psiElement) {
        if(EditorState.getInstance(psiElement.getProject()).showInEditor && psiElement.getFirstChild() == null && psiElement.getText().contains("\n")) {
            GCovCoverageGatherer gatherer = GCovCoverageGatherer.getInstance(psiElement.getProject());
            PsiFile containingFile = psiElement.getContainingFile();
            String path = containingFile.getVirtualFile().getCanonicalPath();
            if(path == null || psiElement.getProject().getBasePath() == null || !path.startsWith(psiElement.getProject().getBasePath())) {
                return null;
            }
            GCovCoverageGatherer.CoverageFileData fileData = gatherer.getCoverageFromPath(path);
            if(fileData == null) {
                return null;
            }

            FileViewProvider fileViewProvider = containingFile.getViewProvider();
            Document document = fileViewProvider.getDocument();
            if(document == null) {
                return null;
            }
            int textOffset = psiElement.getTextOffset();
            int lineNumber = document.getLineNumber(textOffset)+1;
            GCovCoverageGatherer.CoverageLineData lineData =  fileData.getLineData(lineNumber);
            if(lineData == null) {
                return null;
            }

            return new LineMarkerInfo<>(psiElement, new TextRange(textOffset, textOffset), lineData.isCovered() ? CoverageIcons.LINE_COVERED : CoverageIcons.LINE_NOT_COVERED,
                    0, element -> "Executed "  + String.valueOf(lineData.getExecutionCount()) + " times", null, GutterIconRenderer.Alignment.LEFT);
        }
        return null;
    }

    @Override
    public void collectSlowLineMarkers(@NotNull List<PsiElement> list, @NotNull Collection<LineMarkerInfo> collection) { }
}

File: src/gcov/lineMarker/CoverageLineMarker.kt
package gcov.lineMarker

import gcov.Data.CoverageData
import gcov.State.EditorState
import com.intellij.codeInsight.daemon.LineMarkerInfo
import com.intellij.codeInsight.daemon.LineMarkerProvider
import com.intellij.openapi.editor.markup.GutterIconRenderer
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import icons.CoverageIcons

class CoverageLineMarker : LineMarkerProvider {

    override fun getLineMarkerInfo(psiElement: PsiElement): LineMarkerInfo<*>? {
        if (!EditorState.getInstance(psiElement.project).showInEditor || psiElement.firstChild != null || !psiElement.text.contains("\n")) {
            return null
        }

        val gatherer = CoverageData.getInstance(psiElement.project) ?: return null

        val containingFile = psiElement.containingFile
        val path = containingFile.virtualFile.canonicalPath
        if (path == null || psiElement.project.basePath == null || !path.startsWith(psiElement.project.basePath!!)) {
            return null
        }
        val fileData = gatherer.getCoverageFromPath(path) ?: return null

        val fileViewProvider = containingFile.viewProvider
        val document = fileViewProvider.document ?: return null
        val textOffset = psiElement.textOffset
        val lineNumber = document.getLineNumber(textOffset) + 1
        val lineData = fileData.getLineDataAt(lineNumber) ?: return null

        return LineMarkerInfo(psiElement, TextRange(textOffset, textOffset), if (lineData.isCovered) CoverageIcons.LINE_COVERED else CoverageIcons.LINE_NOT_COVERED,
                0, { "Executed " + lineData.executionCount + " times" }, null, GutterIconRenderer.Alignment.LEFT)
    }

    override fun collectSlowLineMarkers(list: List<PsiElement>, collection: Collection<LineMarkerInfo<*>>) {}
}
